packages/core/src/services/documents/destroyOrSoftDeleteDocuments.ts

```diff
@@ -0,0 +1,156 @@
+ * Destroy documents that were NOT CREATED in other commits in the past
+ * This is a hard delete so we don't leave leftovers in the database
+ */
+async function hardDestroyDocuments({
+  documents,
+  existingUuids,
+  tx,
+}: {
+  documents: DocumentVersion[]
+  existingUuids: string[]
+  tx: typeof database
+}) {
+  const uuids = documents
+    .filter((d) => !existingUuids.includes(d.documentUuid))
+    .map((d) => d.documentUuid)
+  if (uuids.length === 0) return
+  return tx
+    .delete(documentVersions)
+    .where(inArray(documentVersions.documentUuid, uuids))
+}
+
+async function createDocumentsAsSoftDeleted({
+  toBeSoftDeleted,
+  commitId,
+  tx,
+}: {
+  toBeSoftDeleted: DocumentVersion[]
+  commitId: number
+  tx: typeof database
+}) {
+  const toBeCreated = toBeSoftDeleted.filter((d) => d.commitId !== commitId)
+
+  if (!toBeCreated.length) return
+
+  return tx.insert(documentVersions).values(
+    toBeCreated.map((d) => ({
+      ...omit(d, ['id', 'updatedAt', 'createdAt']),
+      deletedAt: new Date(),
+      commitId,
+    })),
+  )
+}
+
+async function updateDocumetsAsSoftDeleted({
+  toBeSoftDeleted,
+  commitId,
+  tx,
+}: {
+  toBeSoftDeleted: DocumentVersion[]
+  commitId: number
+  tx: typeof database
+}) {
+  const uuids = toBeSoftDeleted
+    .filter((d) => d.commitId === commitId)
+    .map((d) => d.documentUuid)
+
+  if (!uuids.length) return
+
+  return tx
+    .update(documentVersions)
+    .set({ deletedAt: new Date() })
+    .where(
+      and(
+        inArray(documentVersions.documentUuid, uuids),
+        eq(documentVersions.commitId, commitId),
+      ),
+    )
+}
+
+async function invalidateDocumentsCacheInCommit(
+  commitId: number,
+  tx = database,
+) {
+  return tx
+    .update(documentVersions)
+    .set({ resolvedContent: null })
+    .where(eq(documentVersions.commitId, commitId))
+}
+
+/**
+ * Destroy or soft delete documents in a commit
+ * A document can:
+ *
+ * 1. Not exists in previous commits. In this case, it will be hard deleted
+ * 1. Exists in previous commits and in the commit. It will be updated the `deletedAt` field
+ * 3. Exists in previous commits but not in the commit. It will be created as soft deleted
+ */
+export async function destroyOrSoftDeleteDocuments({
+  documents,
+  commit,
+  trx = database,
+}: {
+  documents: DocumentVersion[]
+  commit: Commit
+  trx?: typeof database
+}): Promise<TypedResult<boolean, Error>> {
+  return Transaction.call(async (tx) => {
+    const commitId = commit.id
+    const existingUuids = await findUuidsInOtherCommits({
+      tx,
+      documents,
+      commitId,
+    })
+    const toBeSoftDeleted = getToBeSoftDeleted({ documents, existingUuids })
+
+    await Promise.all([
+      hardDestroyDocuments({ documents, existingUuids, tx }),
+      createDocumentsAsSoftDeleted({ toBeSoftDeleted, commitId, tx }),
+      updateDocumetsAsSoftDeleted({ toBeSoftDeleted, commitId, tx }),
+      invalidateDocumentsCacheInCommit(commitId, tx),
+    ])
```

Comment:
How is it you're calling both `createDocumentsAsSoftDeleted` and `updateDocumentsAsSoftDeleted` with the same array?

I see that you're filtering them inside each method, but I think it should be more clear to filter it here instead. The `create…` and `update…` services just receive instructions, and if they fail that's responsibility of the caller.

Reply:
I agree. Also the array can be conditionally computed since some of the methods do not require it.